### ISSUE TITLE
Ian todo

### ISSUE THREAD

**Issue by tanguyduval** [neuropoly/qMRLab#39]
implement:
- [x] GenerateButtonsWithPanels
- [x] button_handle2opts
- [x] integrate GenerateButtonsWithPanels in Custom_OptionGUI
- [x] Correct VarNames in all functions
- [x] Change Protocol files in folder data (Create a Parameters.mat for all methods)
- [x] Finish implementing Sim_Single_Voxel_Curve into qMT_SPGR ....
- [x] elargir bouttons de GenerateButtonsWithPanels (hors panels). Add Scale when calling GenerateButtonsWithPanels in CustomOption_GUI
- [x] Delete SfTable if user cancel
- [x] Update function FitGo_FitData(hObject, eventdata, handles) (e.g. remove 'Prot','FitOpt)
- [x] Write a batch for each model
- [x] Add license for all external files
- [x] In B0Mapping Replace FSL prelude to unwarp the Phase With a Matlab function (better than unwarp.m) + add license 
- [x] Put MTV in folder UnderDevelopment. The implementation is very specific to ex vivo dataset only... Many options need to be added to the model.
- [x] In ROI panel, change mouse design ( [set pointer](https://www.mathworks.com/matlabcentral/answers/102914-how-can-i-set-the-mouse-pointer-when-hovering-over-a-certain-uicontrol-or-uipanel-in-matlab-7-7-r20) )to indicate to the user that he needs to draw on the image
- [x] Histogram button should use data within ROI, But should not include the 0 from mask! 
- [x] Statistics and Histogram is redundant.. Show statistics on the histogram plot..
- [x] Remove button Fit Data Menu
- [x] In MWF, cutoff should be in Options panel not in Protocol.
- [x] Add simulation addon for MWF
- [x] Update Model in Simulations (currently you have to open the simulator each time you change in option panel)
- [x] clear Data when method is changed
- [x] Normalize Generatebuttonswithpanel using getpixelposition(ParentHandle)
- [x] Replace all FitDataCustom from all example
- [x] Remove modulaire from data_demo
- [x] remove bug if no starting point
- [x] ROI load not working
- [x] Add test functions that call the batch
- [x] Add Mew "Model" Noise/PIESNO. Ouput sigma + mask
- [x] In "Diffusion" batch. Use PIENO to measure sigma of noise

**Comment by mathieuboudreau** [comment]
@tanguyduval What do you mean by your SPGR_sim question? I glanced at the function, and it doesn't appear that M0r as an input to the function (well, I guess it is by default, through the Sim input). M0r doesn't appear in the GUI as well. What exactly are you asking to be changed here? M0r is used in simulating the signal evolution through the M0 parameter, and appears to be strictly necessary to do so.

**Comment by tanguyduval** [comment]
We saw with Ian that M0r is by default in the DefaultSim.mat
In a commit, we removed M0r because it's confusing to need M0f, F and M0r as input (structure Param).

**Comment by tanguyduval** [comment]
@IanGagnon7, can you please check when done ;-) . qMT methods as classes should be available soon...

**Comment by mathieuboudreau** [comment]
Ok, I was just afraid you were going to refractor the M0 variable in SPGR_sim to not contain the M0r variable, as that type of huge scale refactoring (since the simulation/fitting needs M0r I believe in the 4th array val of M0) could be very sensitive to introducing bugs, particularly since none of the qMT functions are unit tested...

>qMT methods as classes should be available soon...

@tanguyduval @IanGagnon7  Will all the functions in /Model_Functions/SPGR_fun/functions/ remain unchanged after this "qMT method class" addition? The projects that I'm working on that use qMRILab as a submodule are using these functions extensively.

Cheers,

Mathieu

**Comment by tanguyduval** [comment]
Ok thanks, we will add testing functions for the simulations. 
I'm afraid that the refactoring will impact your functions. That wouldn't be clean to have two versions for all qMT methods... We'll contact you when we are ready to integrate your CRLB optimization.

Best,
Tanguy

**Comment by mathieuboudreau** [comment]
Hmm ok, I might need to simply stop contributing/using this version of the repo for a while, and continue my contributions in my of the repo if that's the case. 

I'm too deep in my project to handle an unannounced major modification to features like this.

Cheers,

Mathieu

**Comment by nstikov** [comment]
I am swamped at OHBM, sorry I couldn't act quicker. Obviously we need to make a plan of attack, but I don't understand the problem. The new features are really not a priority, so I'd advise waiting until Mathieu is done whatever it is he is doing. You can always work on another fork.

Starting in July we will have regular meetings to avoid issues like these. 

Learning as I go,

Nikola

**Comment by nstikov** [comment]
By 'you' I meant Tanguy and Ian. What's the time scale of the wait? Hours, days, weeks?

**Comment by mathieuboudreau** [comment]
No worries Nikola; even if the code here was completely changed, I could keep doing my work on my forked repo (I actually have my fork as a submodule, not this one.) or by reverting to a previous commit number. But I like to keep my submodules of qMRI up to date when possible, to be able to push any bug fixes that I might find as I explore it, or improve on the documentation.

With the magic of version control, I could keep working on my project in its current state by simply not pulling the recent commits of this repository, or branching at a commit prior to the changes! =)

@tanguyduval @IanGagnon7 If/when you do make a major change like removing the old qMT functions and replacing it with a class, please tag that commit with a description (and maybe a version number?) so that it can easily be identified. See: https://git-scm.com/book/en/v2/Git-Basics-Tagging and https://stackoverflow.com/questions/5195859/push-a-tag-to-a-remote-repository-using-git

**Comment by tanguyduval** [comment]
Hi everyone,

I'll make a separate branch with Ian modifications.
We didn't change the core qMT functions, we just made cleaner the required inputs (e.g. removed Param.M0r).  We refactored the code by moving some codes, part of qMTLab.m, in independent functions or inside the new qMT classes. 
I know that your CRLB optimization project is massive (I have your OneDrive link). I'll integrate it without modifying your code, just interfacing your inputs/outputs. Is it ready? Do you have an example script? This work will be done on a separate branch.

Best,
Tanguy

**Comment by mathieuboudreau** [comment]
Hi Tanguy,

Don't integrate my crlb code in this repo. I want to wait until my paper is submitted or accepted before making it public, and I will include a link to my own repo in the paper (as well as to qMRLab). 

Afterwards, we can discuss either merging my code in qMRLab, or simply including it as a submodule. The code I sent you is not up to date also, my private repo is the most up to date.

Cheers,

Mathieu

**Comment by tanguyduval** [comment]
Ok I see, So I'll do a local branch for your integration and won't push it. I'll share it with you privatly.

**Comment by mathieuboudreau** [comment]
Ok thanks!

**Comment by IanGagnon7** [comment]
Hi @mathieuboudreau , 

In the SIR-FSE model, Sr value is found by simulation (with the function _computeSr_ which uses Bloch equation), but is it normal that we need to assume values for all theses MT parameters (F, kf, kr, R1f, R1r, T2f, T2r, M0f and lineshape) when Sr should only depends on the lineshape and the T2r according to J.-F. CABANA paper on qMTLab in the SIR-FSE section, "The value of Sr can be determined by numerical simulations and depends on the lineshape of the restricted pool and the assumed value of T2r"?

Best,
Ian

**Comment by mathieuboudreau** [comment]
Hi Ian,

I'm not completely familiar with the SIR-FSE code; is [this](https://github.com/neuropoly/qMRLab/blob/d2196234d31fc938fdf33bf6406df3d39da13940/Models_Functions/SIRFSEfun/functions/computeSr.m) the function you're talking about? 

I don't see this function requiring all the parameters you mentioned. They may be contained in the structure Param, but I can see @jfcabana chose to give then entire Param structure instead of only giving T2r and the lineshape as arguments. The Param structure gives a single container for all parameters, and ensures parameter consistency between function calls. I definitely prefer keeping all the tissue parameters in a single structure, instead of keeping them separate and only giving specific ones as arguments (which makes it much harder to debug later on, and has the potential of causing variable naming conflicts if you're not careful).

Have you encountered a specific issue that giving the entire Param structure as an argument is being problematic for you?

@jfcabana care to chime in?

**Comment by mathieuboudreau** [comment]
+++ Technically, I think that function also uses F implicitely, through Param.M0r (and Param.M0f), since F is Param.M0r/Param.M0f and Param.M0f is set to 1 (I think?).

**Comment by mathieuboudreau** [comment]
Oh wait, I see what you mean for the Bloch simulation line [here](https://github.com/neuropoly/qMRLab/blob/d2196234d31fc938fdf33bf6406df3d39da13940/Models_Functions/SIRFSEfun/functions/computeSr.m#L11).

Hmm, you'll have to ask @jfcabana - I don't know if he still gets email alerts for the GitHub messages here, so it might be useful for you to email him to discuss his implementation of SIR-FSE - I'm not familiar enough to comment more on it.

**Comment by mathieuboudreau** [comment]
I think what JF meant in the lquote you specified is that itdoes depend on those two, but I don't think he meant that it doesnt depend on the other parameters? All MT effects share the coupled equations in the Bloch function I believe, so it makes sense to me at first glance that these simulations would require those parameters as well.

**Comment by IanGagnon7** [comment]
I think you just found it, all the parameters I enumerate you are needed to calculate the magnetization vector [Mxf, Myf, Mzf, Mzr] called M_temp in the function. So we needed them all to compute the Sr!
And I found the proof in qMTLab paper: "Model parameters available for fitting are: F, krf, R1f, R1r, Sf, Sr, and M0f. The starting (or fixed) value of Sr can be computed directly in the interface, by **setting the assumed mean MT parameters** (T2r in particular), and clicking on “Compute Sr.” This will perform a simulation of the inversion pulse, and compute the fraction of the restricted pool magnetization after vs. before the pulse. This value is strongly dependent on the lineshape of the restricted pool and on T2r."
Thank you @mathieuboudreau!

**Comment by mathieuboudreau** [comment]
Great! - feel free to message me for anything else!

Good luck!

**Comment by jfcabana** [comment]
Hello @IanGagnon7 and @mathieuboudreau ,

Yes you pretty much figured this out correctly :) 

The computation of Sr is based on the full Bloch equations, and thus is dependant on all parameters. But the ones that are the more important are the lineshape of the restricted pool and T2r. Sr is still dependant on the other parameters, but since the inversion pulse is of very short duration, not much MT or T1 recovery is going on during that time. 

From what I remember, I think in the original SIR-FSE method paper from Gochberg, they assume no MT or T1 recovery during the inversion pulse for the computation of Sr, but I'm not sure. I just thought it made more sense to include it, based on average values of parameters that could be in the ballpark of the tissue you are measuring. Anyway, I think the effect is very small on Sr, but not zero. You could try it in the software by varying MT parameters to see the effect on Sr...

And thanks guys for all the work you are putting in this! I moved on to other projects, not even involving MRI now, so I kind of have to let my baby grow up in the hands of other experts now :)

Cheers!

**Comment by tanguyduval** [comment]
Hi guys,

One problem is still present..
Why do you force Sf values to be positive? https://github.com/neuropoly/qMRLab/blob/master/Models_Functions/SIRFSEfun/functions/SIRFSE_fit.m#L49
It prevents good fitting because Sf is usually negative:
**without absolute:**
![image](https://user-images.githubusercontent.com/7785316/28284860-816c4fc8-6b00-11e7-9b64-eaf965a0d4c4.png)
**with absolute:**
![image](https://user-images.githubusercontent.com/7785316/28284876-9480625c-6b00-11e7-9864-2ecac08d61e1.png)

Thanks,
Tanguy

**Comment by mathieuboudreau** [comment]
@jfcabana will be better at answering this one, I'm not sure of the answer.

**Comment by mathieuboudreau** [comment]
@tanguyduval It's odd because in JF's qMTLab paper, Figure 5 shows a dataset similar to yours and the fit is great.

And if you look at "Blame" in the function you linked, he hasn't changed that line since the paper was published.

Did you test to see if this "bug"/bad fit is present in a commit closer to the paper submission (e.g. https://github.com/neuropoly/qMTLab)? If it worked then and not now, I doubt that the abs() is the bug, and likely more a bug introduced elsewhere since then.

**Comment by jfcabana** [comment]
Hi guys,

Sf is indeed a negative value. I can't remember exactly why I introduced the abs() in the Fit struct, but I know it was a quick workaround to fix some bug... The code is working fine in my implementation of qMTLab as it was when the paper was published. Maybe something changed in the way the curve gets computed/plotted?

Anyways, the abs() doesn't affect the fitting, as it is applied only after the fit is done. It only affect displayed curve. You could remove it if you want/need to, as long as results are consistent. If you do so, you would also need to change line 6 [here ](https://github.com/neuropoly/qMRLab/blob/master/Models_Functions/SIRFSEfun/functions/SIRFSE_SimCurve.m) :
xFit = [Fit.F, Fit.kr, Fit.R1f, Fit.R1r, -1*Fit.Sf, Fit.Sr, Fit.M0f]; -> xFit = [Fit.F, Fit.kr, Fit.R1f, Fit.R1r, Fit.Sf, Fit.Sr, Fit.M0f];

Cheers!

**Comment by tanguyduval** [comment]
Ok thank you Mathieu and JF, I didn't noticed that it was done after the fitting...

Both the original SIRFSE method and the new one are working properly. I corrected this in d6513b454dd70c76c96fd2546a01a475132f1d0a

**Comment by mathieuboudreau** [comment]
Great! A+

**Comment by tanguyduval** [comment]
@IanGagnon7 All script in External come from FileExchange. You don't need to write an email to each author, but you need to create a folder per file with the license file attached. 
To get the license, go on https://www.mathworks.com/matlabcentral/fileexchange/, search for the script, and click on view license.. It's the same license all the time but author and date change (Matlab's FileExchange has a unique copyright)

**Comment by mathieuboudreau** [comment]
@tanguyduval I just checked, and it appears that NODDI uses the Open Source-approved Artistic Licence 2.0, which means we could be including it directly in our own repo ([tl;dr of the license](https://tldrlegal.com/license/artistic-license-2.0-(artistic)), [full license explanation](https://opensource.org/licenses/Artistic-2.0)). 

Is there a reason why we don't? Is it too large, or does it change frequently?

**Comment by tanguyduval** [comment]
done in f8012ee04b4357e50d833044108e5049b501430e
Thanks!

**Comment by mathieuboudreau** [comment]
Ah ha! Nice! I had listed NODDI as an external dependency in the README file (because I had to download it myself to get it working previously). I think it could be removed now right?

**Comment by tanguyduval** [comment]
Oh yes you can :-)

**Comment by tanguyduval** [comment]
@IanGagnon7 I modified some of your tasks (notably qMRLab.fig ), so pull to prevent conflicts!

**Comment by mathieuboudreau** [comment]
@IanGagnon7 @tanguyduval FYI, I increased the upperbound for kr in SPGR modulaire from 30 to 500 in [this commit](https://github.com/neuropoly/qMRLab/commit/0d7caadfff7194e73dfc2209aa128ee428714f42#diff-431d209f18238ead17447924ced30116L43) today.

kr's typical values are actually around 30 (see Sled et al's 2001 paper, or any other SPGR qMT papers), so an upper bound this low is too short of a limit, and also propagates errors to F and R1f (as they are all closely related). I noticed this insufficient limit this morning while doing some Monte Carlo on an old commit - I think JF might have initially used that limit for kf (which has values typically around 5), but forgot to update the up when he switched from kf->kr.

For low SNR values, kf's upperbound could even be upped a bit more, as it's a very unstable parameter, however 500 sounded like a good typical UB for me for most cases.

**Comment by mathieuboudreau** [comment]
Also, when this branch finally does get merged, can we do it through a Github pull reques,t instead of simply "git merge", and leave the pull request up for a few days? It would be best if we could take a look at all the file changes before fully merging, to ensure that no file was deleted accidentally and lost/forgotten in this major update.

**Comment by nstikov** [comment]
Happy to see this effort evolve organically!

At some point at the end of summer I would like us to start with bi-weekly
meetings where we come up with a feasible long-term strategy. I am back
from e-mail hiatus, so feel free to follow up with me.

**Comment by IanGagnon7** [comment]
@mathieuboudreau  I am ready to make the pull request of my branch, I would only have to remove the "old" qMT models and the word "modulaire" from the "new" qMT models to make them official! Is that ok for you Mathieu? Let me know if you would prefer to wait a little bit or if there are things you'd like me to change. 

Best,
Ian

**Comment by mathieuboudreau** [comment]
@IanGagnon7 Sure go ahead! I'm working off my fork right now anyways, so if it's buggy it won't affect my stuff.

Cheers,

Mathieu

**Comment by mathieuboudreau** [comment]
@IanGagnon7 After your remove the "old" qMT Models and everything else needed, but before you push the branch, try and run all the tests to make sure that they all pass (there's not whole-code coverage yet, but at least some coverage).

You can do it by running this command from the qMRLab repository folder in Matlab: `result = runtests(pwd, 'Recursively', true)`

**Comment by tanguyduval** [comment]
Add acknoledgements for the core functions.
For instance in MWF:
````
`
% ----------------------------------------------------------------------------------------------------
% Written by: Ian Gagnon, 2017
% Core functions written by: Eva Alonso Ortiz, 20??
% Reference: FILL
% ----------------------------------------------------------------------------------------------------
`````

Also ask Tung for the references!

**Comment by mathieuboudreau** [comment]
@tanguyduval writing acknowledgments for all the core functions (e.g. Written by: ) would get messy and inaccurate for a while, no? Especially since we are keeping the package open sourced, and anyone can contribute/change the files. It can be easy to lose track over a while as well, no?

Isn't the repository license and git history sufficient? The authors of the files are stored in the git history anyhow. Or maybe we could add a generic "Copyright Neuropoly" to all the files instead?

**Comment by mathieuboudreau** [comment]
Adding references to the original papers/equations would be great though.

**Comment by nstikov** [comment]
Agree with Matt, reference the paper, as these fits have been passed on for
generations

P.S. Sent from a cell phone, please excuse the terseness.

**Comment by tanguyduval** [comment]
@IanGagnon7 good job!
